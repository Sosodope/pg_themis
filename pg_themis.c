#include "postgres.h"
#include "fmgr.h"
#include <stdlib.h>
#include <themis/themis.h>

PG_MODULE_MAGIC;

PG_FUNCTION_INFO_V1(pg_themis_scell_encrypt_seal);
PG_FUNCTION_INFO_V1(pg_themis_scell_decrypt_seal);
PG_FUNCTION_INFO_V1(pg_themis_smessage_encrypt);
PG_FUNCTION_INFO_V1(pg_themis_smessage_decrypt);

Datum pg_themis_scell_encrypt_seal(PG_FUNCTION_ARGS){

  bytea* data = PG_GETARG_BYTEA_P(0);
  bytea* key = PG_GETARG_BYTEA_P(1);
  size_t encrypted_data_len=0;
  bytea* encrypted_data=NULL;

  if(THEMIS_BUFFER_TOO_SMALL!=themis_secure_cell_encrypt_seal((const uint8_t*)VARDATA(key), VARSIZE(key)-VARHDRSZ, NULL, 0, (const uint8_t*)VARDATA(data), VARSIZE(data)-VARHDRSZ, NULL, &encrypted_data_len)){
    ereport(ERROR,(errcode(-1), errmsg("ERR secure seal encription failed"), errdetail("length determination"), NULL));
  }
  encrypted_data = palloc(encrypted_data_len+VARHDRSZ);
  if(!encrypted_data){
    ereport(ERROR,(errcode(-1), errmsg("ERR allocation error"), NULL, NULL));
  }
  SET_VARSIZE(encrypted_data, encrypted_data_len+VARHDRSZ);
  if(THEMIS_SUCCESS!=themis_secure_cell_encrypt_seal((const uint8_t*)VARDATA(key), VARSIZE(key)-VARHDRSZ, NULL, 0, (const uint8_t*)VARDATA(data), VARSIZE(data)-VARHDRSZ, (uint8_t*)VARDATA(encrypted_data), &encrypted_data_len)){
    pfree(encrypted_data);
    ereport(ERROR,(errcode(-1), errmsg("ERR secure seal encription failed"), NULL, NULL));
  }
  PG_RETURN_BYTEA_P(encrypted_data);
}

Datum pg_themis_scell_decrypt_seal(PG_FUNCTION_ARGS){

  bytea* data = PG_GETARG_BYTEA_P(0);
  bytea* key = PG_GETARG_BYTEA_P(1);
  size_t decrypted_data_len=0;
  bytea* decrypted_data=NULL;

  if(THEMIS_BUFFER_TOO_SMALL!=themis_secure_cell_decrypt_seal((const uint8_t*)VARDATA(key), VARSIZE(key)-VARHDRSZ, NULL, 0, (const uint8_t*)VARDATA(data), VARSIZE(data)-VARHDRSZ, NULL, &decrypted_data_len)){
    ereport(ERROR,(errcode(-1), errmsg("ERR secure seal decription failed"), errdetail("length determination"), NULL));
  }
  decrypted_data = palloc(decrypted_data_len+VARHDRSZ);
  if(!decrypted_data){
    ereport(ERROR,(errcode(-1), errmsg("ERR allocation error"), NULL, NULL));
  }
  SET_VARSIZE(decrypted_data, decrypted_data_len+VARHDRSZ);
  if(THEMIS_SUCCESS!=themis_secure_cell_decrypt_seal((const uint8_t*)VARDATA(key), VARSIZE(key)-VARHDRSZ, NULL, 0, (const uint8_t*)VARDATA(data), VARSIZE(data)-VARHDRSZ, (uint8_t*)VARDATA(decrypted_data), &decrypted_data_len)){
    pfree(decrypted_data);
    ereport(ERROR,(errcode(-1), errmsg("ERR secure seal decription failed"), NULL, NULL));
  }
  PG_RETURN_BYTEA_P(decrypted_data);
}

//encrypt data with acra ctruct 
static int smessage_encrypt(const uint8_t* data, const uint32_t data_length, const uint8_t* private_key, const uint32_t private_key_length, const uint8_t* public_key, const uint32_t public_key_length, const uint8_t* peer_public_key, const uint32_t peer_public_key_length, bytea** enc_data){
  size_t encrypted_data_len=0;
  if(THEMIS_BUFFER_TOO_SMALL!=themis_secure_message_wrap(private_key, private_key_length, peer_public_key, peer_public_key_length, data, data_length, NULL, &encrypted_data_len)){
    return -1;
  }
  *enc_data = palloc(public_key_length+sizeof(uint32_t)+encrypted_data_len+VARHDRSZ);
  if(!(*enc_data)){
    return -2;
  }
  memcpy((uint8_t*)VARDATA(*enc_data), &public_key_length, sizeof(public_key_length));
  memcpy((uint8_t*)VARDATA(*enc_data)+sizeof(public_key_length), public_key, public_key_length);
  if(THEMIS_SUCCESS!=themis_secure_message_wrap(private_key, private_key_length, peer_public_key, peer_public_key_length, data, data_length, (uint8_t*)VARDATA(*enc_data)+sizeof(public_key_length)+public_key_length, &encrypted_data_len)){
    pfree(*enc_data);
    return -1;
  }
  encrypted_data_len+=sizeof(public_key_length)+public_key_length;
  SET_VARSIZE(*enc_data, encrypted_data_len+VARHDRSZ);
  return 0;
}

//decrypt  acra structed data 
static int smessage_decrypt(const uint8_t* data, const uint32_t data_length, const uint8_t* private_key, const uint32_t private_key_length, bytea** dec_data){
  size_t decrypted_data_len=0;
  const uint32_t public_key_length = *((const uint32_t*)data);
  const uint8_t* public_key = data+sizeof(uint32_t);
  const uint8_t* data_ = public_key+public_key_length;
  size_t data_length_ = data_length-public_key_length-sizeof(uint32_t); 
  if(data_length<=public_key_length+sizeof(uint32_t)){
    return -1;
  }
  
  if(THEMIS_BUFFER_TOO_SMALL!=themis_secure_message_unwrap(private_key, private_key_length, public_key, public_key_length, data_, data_length_, NULL, &decrypted_data_len)){
    return -1;
  }
  *dec_data = palloc(decrypted_data_len+VARHDRSZ);
  if(!(*dec_data)){
    return -2;
  }
  if(THEMIS_SUCCESS!=themis_secure_message_unwrap(private_key, private_key_length, public_key, public_key_length, data_, data_length_, (uint8_t*)VARDATA(*dec_data), &decrypted_data_len)){
    pfree(*dec_data);
    return -1;
  }
  SET_VARSIZE(*dec_data, decrypted_data_len+VARHDRSZ);
  return 0;
}

static int smessage_enc(const uint8_t* public_key, const uint32_t public_key_length, const uint8_t* data, const uint8_t data_length, bytea** enc_data){
  uint8_t new_private_key[10240];
  uint8_t new_public_key[10240];

  size_t new_private_key_length=10240, new_public_key_length=10240;
  if(THEMIS_SUCCESS!=themis_gen_ec_key_pair(new_private_key, &new_private_key_length, new_public_key, &new_public_key_length)){
    return -1;
  }
  return smessage_encrypt(data, data_length, new_private_key, new_private_key_length, new_public_key, new_public_key_length, public_key, public_key_length, enc_data);
}

static int smessage_dec(const uint8_t* private_key, const uint32_t private_key_length, const uint8_t* data, const uint8_t data_length, bytea** dec_data){
  return smessage_decrypt(data, data_length, private_key, private_key_length, dec_data);
}

Datum pg_themis_smessage_encrypt(PG_FUNCTION_ARGS){
  bytea* data = PG_GETARG_BYTEA_P(0);
  bytea* key = PG_GETARG_BYTEA_P(1);
  bytea* encrypted_data=NULL;

  if(0!=smessage_enc((const uint8_t*)VARDATA(key), VARSIZE(key)-VARHDRSZ, (const uint8_t*)VARDATA(data), VARSIZE(data)-VARHDRSZ, &encrypted_data)){
    ereport(ERROR,(errcode(-1), errmsg("ERR secure message encription failed"), NULL, NULL));
  }
  PG_RETURN_BYTEA_P(encrypted_data);
}
Datum pg_themis_smessage_decrypt(PG_FUNCTION_ARGS){
  bytea* data = PG_GETARG_BYTEA_P(0);
  bytea* key = PG_GETARG_BYTEA_P(1);
  bytea* decrypted_data=NULL;

  if(0!=smessage_dec((const uint8_t*)VARDATA(key), VARSIZE(key)-VARHDRSZ, (const uint8_t*)VARDATA(data), VARSIZE(data)-VARHDRSZ, &decrypted_data)){
    ereport(ERROR,(errcode(-1), errmsg("ERR secure message decription failed"), NULL, NULL));
  }
  PG_RETURN_BYTEA_P(decrypted_data);
}
