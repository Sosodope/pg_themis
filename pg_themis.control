comment = 'pg_themis'
default_version = '1.0'

relocatable = true
