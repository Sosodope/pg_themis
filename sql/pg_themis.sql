CREATE EXTENSION themis;

select pg_themis_scell_decrypt_seal(pg_themis_scell_encrypt_seal(E'test data', E'test_password'), E'test_password');

select pg_themis_smessage_decrypt(pg_themis_smessage_encrypt(E'test data', '\x554543320000002d6bbb797903fab7333a4d6eb7c259de7896fa69e6638691c265a092f65a223ca98ec9a73542'), '\x524543320000002dc7a8ca7a00c3b5d1ad5137308f45e65e54df2b7a45bc8508e8cc3bc9481b631ae8128b3974');

DROP EXTENSION themis;
