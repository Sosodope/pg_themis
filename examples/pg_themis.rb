#!/usr/local/bin/ruby
#
# Copyright (c) 2016 Cossack Labs Limited
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require "pg"
require "rubythemis"

conn = PG.connect( :dbname => 'test' )

# load module 
conn.exec("CREATE EXTENSION IF NOT EXISTS pg_themis")

#scell encrypt with pg_themis
conn.prepare('stmt', "select pg_themis_scell_encrypt_seal($1::bytea, $2::bytea)")
enc_data = conn.unescape_bytea(conn.exec_prepared('stmt', ["data", "password"]).values[0][0])
p enc_data

#scell decrypt by rubythemis
seal=Themis::Scell.new("password", Themis::Scell::SEAL_MODE)
data=seal.decrypt(enc_data)
p data

#scell encrypt by rubythemis
enc_data = seal.encrypt("data")
p enc_data

#seal decrypt by pg_themis
conn.prepare('stmt1', "select pg_themis_scell_decrypt_seal($1::bytea, $2::bytea)")
data = conn.unescape_bytea(conn.exec_prepared('stmt1', [conn.escape_bytea(enc_data), "password"]).values[0][0])
p data

#smessage encrypt
conn.prepare('stmt2', "select pg_themis_smessage_encrypt($1::bytea, $2::bytea)")
enc_data = conn.unescape_bytea(conn.exec_prepared('stmt2', ["data", conn.escape_bytea("\x55\x45\x43\x32\x00\x00\x00\x2d\x6b\xbb\x79\x79\x03\xfa\xb7\x33\x3a\x4d\x6e\xb7\xc2\x59\xde\x78\x96\xfa\x69\xe6\x63\x86\x91\xc2\x65\xa0\x92\xf6\x5a\x22\x3c\xa9\x8e\xc9\xa7\x35\x42")]).values[0][0])
p enc_data

#smessage decrypt
conn.prepare('stmt3', "select pg_themis_smessage_decrypt($1::bytea, $2::bytea)")
data = conn.unescape_bytea(conn.exec_prepared('stmt3', [conn.escape_bytea(enc_data), conn.escape_bytea("\x52\x45\x43\x32\x00\x00\x00\x2d\xc7\xa8\xca\x7a\x00\xc3\xb5\xd1\xad\x51\x37\x30\x8f\x45\xe6\x5e\x54\xdf\x2b\x7a\x45\xbc\x85\x08\xe8\xcc\x3b\xc9\x48\x1b\x63\x1a\xe8\x12\x8b\x39\x74")]).values[0][0])
p data

# unload module
conn.exec("DROP EXTENSION pg_themis")

conn.close if conn