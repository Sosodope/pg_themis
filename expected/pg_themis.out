CREATE EXTENSION
 pg_themis_scell_decrypt_seal 
------------------------------
 \x746573742064617461
(1 row)

 pg_themis_smessage_decrypt 
----------------------------
 \x746573742064617461
(1 row)

DROP EXTENSION
